### DISCUSSION TITLE
Expiry and tokenKeys of `IHederaTokenService.HederaToken`

### DISCUSSION THREAD

**Discussion by daniel-rekursive** [hashgraph/hedera-smart-contracts#103] [Q&A]
`HederaTokenService.createFungibleToken()` function accepts `IHederaTokenService.HederaToken` as its first arg.

```Solidity
struct HederaToken {
  // The publicly visible name of the token. The token name is specified as a Unicode string.
  // Its UTF-8 encoding cannot exceed 100 bytes, and cannot contain the 0 byte (NUL).
  string name;
  // The publicly visible token symbol. The token symbol is specified as a Unicode string.
  // Its UTF-8 encoding cannot exceed 100 bytes, and cannot contain the 0 byte (NUL).
  string symbol;
  // The ID of the account which will act as a treasury for the token as a solidity address.
  // This account will receive the specified initial supply or the newly minted NFTs in
  // the case for NON_FUNGIBLE_UNIQUE Type
  address treasury;
  // The memo associated with the token (UTF-8 encoding max 100 bytes)
  string memo;
  // IWA compatibility. Specified the token supply type. Defaults to INFINITE
  bool tokenSupplyType;
  // IWA Compatibility. Depends on TokenSupplyType. For tokens of type FUNGIBLE_COMMON - the
  // maximum number of tokens that can be in circulation. For tokens of type NON_FUNGIBLE_UNIQUE -
  // the maximum number of NFTs (serial numbers) that can be minted. This field can never be changed!
  uint32 maxSupply;
  // The default Freeze status (frozen or unfrozen) of Hedera accounts relative to this token. If
  // true, an account must be unfrozen before it can receive the token
  bool freezeDefault;
  // list of keys to set to the token
  TokenKey[] tokenKeys;
  // expiry properties of a Hedera token - second, autoRenewAccount, autoRenewPeriod
  Expiry expiry;
}
```
- What value should `tokenKeys` and `expiry` be? Any detailed doc describing them?
- Is it okay to call `HederaTokenService.createFungibleToken(token, initialTotalSupply, decimals)` where `token.tokenKeys` and `token.expiry` are empty?

Example code here doesn't give me clear understanding of what those values are and what it should accept.
https://github.com/hashgraph/hedera-smart-contracts/blob/main/hts-precompile/TokenCreateContract.sol#L44-L46

**Comment by daniel-rekursive** [comment]
cc @shemnon @SimiHunjan @dimitar-dinev 
Thanks

**Comment by stoqnkpL** [comment]
Hey @daniel-rekursive,

`TokenKey` and `Expiry` are struct types defined in [IHederaTokenService](https://github.com/hashgraph/hedera-smart-contracts/blob/decd796ab1a51fdd5a4c729cb5aaf7498ea2ff7d/hts-precompile/IHederaTokenService.sol#L53). We have an [open PR](https://github.com/hashgraph/hedera-improvement-proposal/pull/448) with updates to HIP-358 that provide detailed explanations of all new structs and how to use them as arguments when creating tokens. Please take a look at the description of different fields that's provided there, hopefully it would make things clearer!
As for your second question, the rule of thumb is that the same restrictions and requirements for token creation apply when creating a token through the precompile as would apply if the same token is created through HTS HAPI.

**Comment by daniel-rekursive** [comment]
For the second question, does it mean it's okay to leave `token.tokenKeys` and `token.expiry` without assigning any value?
@stoqnkpL

**Comment by stoqnkpL** [comment]
Hey @daniel-rekursive,
It is okay to leave `tokenKeys` empty, but keep in mind that a token without an admin key is immutable and cannot be updated.
For the expiry there are two valid options:
1. Auto renew account is provided and the auto renew period is valid
2. The second that was passed (IHederaTokenService.Expiry.second) is in the future compared to token creation time

All other cases would result in either `INVALID_RENEWAL_PERIOD` or `INVALID_EXPIRATION_TIME` status and token creation will fail.

**Comment by daniel-rekursive** [comment]
>token without an admin key is immutable and cannot be updated

@stoqnkpL  What do you mean it is "immutable" and "cannot be updated"?

**Comment by stoqnkpL** [comment]
Usually, HTS Tokens can be updated, for example, the token name, symbol, treasury account, or other token properties can be changed on an existing token. In order to be able to execute a token update, however, you would need to sign the transaction with the token admin key. If the token has no admin key, this means it can never be updated or deleted.

**Comment by stoqnkpL** [comment]
Hey Daniel,
Sorry for the delayed response. Is this issue still relevant? I'm trying to
reproduce it, can you share a bit more context, how are you constructing
the arguments for createFungileToken()?

**Comment by Kalina-Todorova** [comment]
Hi @daniel-rekursive is this still an issue for you?

**Comment by Nana-EC** [comment]
@daniel-rekursive did you receive the information you needed.
Notably this should be a discussion item so i'll be converting this over
